### PR TITLE
feat: handle VRF Gateways in "metal gateway list"

### PR DESCRIPTION
**Pull request by displague**
Adds `metal gateway list` support for VRF Gateways.

**Why is this needed?**

```
$ metal gateway get
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xad902e]

goroutine 1 [running]:
github.com/equinix/metal-cli/internal/gateway.(*Client).Retrieve.func1(0xc00021c308?, {0x12e08e0?, 0x0?, 0x0?})
        /home/runner/work/metal-cli/metal-cli/internal/gateway/retrieve.go:70 +0x56e
github.com/spf13/cobra.(*Command).execute(0xc00021c308, {0x12e08e0, 0x0, 0x0})
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:983 +0xaca
github.com/spf13/cobra.(*Command).ExecuteC(0xc000134608)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(0xc0000061c0?)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039 +0x13
main.main()
        /home/runner/work/metal-cli/metal-cli/cmd/metal/main.go:32 +0x1b
```

### CODE REVIEW

**Comment by displague** [comment]
```
metal-cli % PACKNGO_DEBUG=1 go run ./cmd/metal/ gateways list                              
GET /metal/v1//projects/{project_id}/metal-gateways?include=virtual_network%2Cip_reservation HTTP/1.1
Host: api.equinix.com
User-Agent: metal-cli/(devel) equinix-sdk-go/0.42.0
Accept: application/json
X-Auth-Token: **REDACTED**
Accept-Encoding: gzip

HTTP/1.1 404 Not Found
Content-Type: application/json; charset=utf-8

{"errors":["Not found"]}
Error: Could not list Project Metal Gateways: 404 Not Found Not found
```

**Comment by displague** [comment]
My default project ID was old (no longer exists).

Updated to a valid one:
```console
go run ./cmd/metal/ gateways list -p $METAL_PROJECT_ID 
+---------------+-------+-------+------------------+--------+-------------------------------+-----------------+
|         ID    | METRO | VXLAN |    ADDRESSES     | STATE  |            CREATED            |      TYPE       |
+---------------+-------+-------+------------------+--------+-------------------------------+-----------------+
| redacted-uuid | am    | 1002  | 192.168.100.2/22 | active | 2024-08-02 18:16:23 +0000 UTC | VrfMetalGateway |
+---------------+-------+-------+------------------+--------+-------------------------------+-----------------+
```

File: internal/gateway/retrieve.go [hunk 1]
```diff
@@ -24,12 +24,67 @@ import (
...
+
+func (m metalGatewayWrapper) GetIpReservation() ipreservationIsh {
+	if m.MetalGateway.IpReservation == nil {
+		return nil
+	}
+	return m.MetalGateway.IpReservation
+}
+
+func (m metalGatewayWrapper) GatewayType() string {
+	return "MetalGateway"
```
Comment (displague) [inline internal/gateway/retrieve.go]: Is this name ok?

Comment (ctreatma) [reply internal/gateway/retrieve.go]: Other customer-facing tools (web console, terraform, Ansible, etc.) don't expose a type for gateways so I don't think we should introduce one here.  Terraform exposes a `vrf_id` attribute that is taken from the IP reservation, and we could emulate that in the text output for gateway commands.

File: internal/gateway/retrieve.go [hunk 2]
```diff
@@ -58,26 +113,45 @@ func (c *Client) Retrieve() *cobra.Command {
...
 			gways := gwayList.GetMetalGateways()
 
-			data := make([][]string, len(gways))
-			metalGways := make([]*metal.MetalGateway, len(gways))
+			data := make([][]string, 0, len(gways))
+			var gateways []gatewayIsh
+
+			for _, gwaysInner := range gways {
+				var gway gatewayIsh
+				if gwaysInner.MetalGateway != nil {
```
Comment (ctreatma) [inline internal/gateway/retrieve.go]: IMO introducing additional types/interfaces isn't worth it if we're still going to end up with an if block inside a for loop.

File: internal/gateway/retrieve.go (hunk 1 above)
Comment (displague) [reply internal/gateway/retrieve.go]: I didn't find examples of other IDs being included in the output, so I included VRF Name. The meant adding "vrf" to the `include` list as only the `href` was available.

File: internal/gateway/retrieve.go (hunk 2 above)
Comment (displague) [reply internal/gateway/retrieve.go]: I moved this out of the loop

**Comment by displague** [comment]
Output with a VRF and non-VRF Gateway:
```
$ go run ./cmd/metal/ gateways list -p $METAL_PROJECT_ID
+--------------------------------------+-------+-------+------------------+--------+-------------------------------+-------------------+
|                  ID                  | METRO | VXLAN |    ADDRESSES     | STATE  |            CREATED            |        VRF        |
+--------------------------------------+-------+-------+------------------+--------+-------------------------------+-------------------+
| 48492792-aa66-4857-aac3-abcd12345678 | am    | 1000  | 10.12.0.242/29   | ready  | 2024-08-08 18:44:48 +0000 UTC |                   |
| bd3162f4-4162-408b-8670-abcd12345678 | am    | 1002  | 192.168.100.2/22 | active | 2024-08-02 18:16:23 +0000 UTC | nutanix-vrf-bRkYc |
+--------------------------------------+-------+-------+------------------+--------+-------------------------------+-------------------+
```

File: internal/gateway/retrieve.go [hunk 3]
```diff
@@ -58,26 +124,41 @@ func (c *Client) Retrieve() *cobra.Command {
...
-			data := make([][]string, len(gways))
-			metalGways := make([]*metal.MetalGateway, len(gways))
+			data := make([][]string, 0, len(gways))
+			var gateways []gatewayIsh
 
-			for i, n := range gways {
-				gway := n.MetalGateway
-				metalGways = append(metalGways, gway)
+			for _, gwaysInner := range gways {
+				gway := newGatewayish(gwaysInner)
```
Comment (ctreatma) [inline internal/gateway/retrieve.go:131]: Does this _really_ count as getting rid of the if block within the for loop?

**Review by ctreatma — APPROVED** [review]

**Pull request merged**